src/include/index/flatpq_index.h

```diff
@@ -0,0 +1,927 @@
+
+#include <tiledb/group_experimental.h>
+#include <tiledb/tiledb>
+
+template <class T, class U>
+void sub_kmeans_random_init(
+    const ColMajorMatrix<T>& training_set,
+    ColMajorMatrix<U>& centroids,
+    size_t sub_begin,
+    size_t sub_end,
+    size_t seed = 0) {
+  scoped_timer _{__FUNCTION__};
+
+  if (num_vectors(training_set) < num_vectors(centroids)) {
+    throw std::invalid_argument(
+        "Number of vectors in training set must be greater than or equal to "
+        "number of centroids");
+  }
+
+  std::mt19937 gen(seed == 0 ? std::random_device{}() : seed);
+  std::uniform_int_distribution<> dis(0, num_vectors(training_set) - 1);
+
+  size_t num_clusters = num_vectors(centroids);
+
+  std::vector<size_t> indices(num_clusters);
+  std::unordered_set<size_t> visited;
+  for (size_t i = 0; i < num_clusters; ++i) {
+    size_t index;
+    do {
+      index = dis(gen);
+    } while (visited.contains(index));
+    indices[i] = index;
+    visited.insert(index);
+  }
+  // std::iota(begin(indices), end(indices), 0);
+  // std::shuffle(begin(indices), end(indices), gen);
+
+  for (size_t i = 0; i < num_clusters; ++i) {
+    for (size_t j = sub_begin; j < sub_end; ++j) {
+      centroids(j, i) = training_set(j, indices[i]);
+    }
+  }
+}
+
+// #define REASSIGN
+
+/**
+ * @brief Run kmeans on a subspace
+ * @tparam T Data type of the input vectors
+ * @param training_set Training set
+ * @param centroids Initial locations of centroids.  Will be updated in
+ * locations [sub_begin, sub_end).
+ * @param sub_begin Beginning of the subspace
+ * @param sub_end End of the subspace
+ * @param num_clusters Number of clusters to find
+ * @return
+ *
+ * @todo update with concepts
+ * @todo fix up zero sized partitions
+ * @todo this would be more cache friendly to do sub_begin and sub_end
+ * in an inner loop
+ * @todo We can probably just reuse plain kmeans at some point.
+ */
+template <class T, class U>
+auto sub_kmeans(
+    const ColMajorMatrix<T>& training_set,
+    ColMajorMatrix<U>& centroids,
+    size_t sub_begin,
+    size_t sub_end,
+    size_t num_clusters,
+    double tol,
+    size_t max_iter,
+    size_t num_threads,
+    float reassign_ratio = 0.05) {
+  // using feature_type = T;
+  // using centroids_feature_type = U;
+  // using index_type = size_t;
+
+  size_t sub_dimension_ = sub_end - sub_begin;
+
+  std::vector<size_t> degrees(num_clusters, 0);
+
+  // Copy centroids to new centroids -- note only one subspace will be changing
+  // @todo Keep new_centroids outside function so we don't need to copy all
+  ColMajorMatrix<U> new_centroids(dimension(centroids), num_vectors(centroids));
+  for (size_t i = 0; i < num_vectors(new_centroids); ++i) {
+    for (size_t j = 0; j < dimension(new_centroids); ++j) {
+      new_centroids(j, i) = centroids(j, i);
+    }
+  }
+
+  size_t iter = 0;
+  double max_diff = 0.0;
+  double total_weight = 0.0;
+  for (iter = 0; iter < max_iter; ++iter) {
+#ifdef REASSIGN
```

Comment:
Should this be a function argument rather than #ifdef?

Reply:
Yes, although we can't directly substitute an `if` statement for the `ifdef` as there are some structural / scoping issues that have to be handled.  Deferring to future PR (sc-40508).